File: src/core/dead-link-detector.ts
// Dead Link Detector — Pure functions for dead link detection and correction
// Extracted from lint-fixes.ts::fixDeadLink()
// Zero side effects, fully testable

export interface PageRef {
  path: string;
  title: string;
  aliases?: string[];
}

/**
 * Find matching page for a dead link target.
 * Searches by title (case-insensitive) first, then by aliases.
 *
 * @param pages - Available wiki pages from getExistingWikiPages
 * @param targetName - The dead link target (e.g., "思维链" or "CoT")
 * @returns Matching page or undefined if no match found
 *
 * @example
 * const pages = [{ title: 'Chain of Thought', aliases: ['CoT', '思维链'] }];
 * findDeadLinkTarget(pages, '思维链') // => { title: 'Chain of Thought', ... }
 * findDeadLinkTarget(pages, 'cot') // => { title: 'Chain of Thought', ... }
 * findDeadLinkTarget(pages, 'nonexistent') // => undefined
 */
export function findDeadLinkTarget(
  pages: PageRef[],
  targetName: string
): PageRef | undefined {
  const targetBasename = targetName.includes('/')
    ? targetName.split('/').pop()!
    : targetName;
  const lowerTarget = targetBasename.toLowerCase();

  // First: exact title match (case-insensitive)
  let match = pages.find(p => p.title.toLowerCase() === lowerTarget);

  // Second: alias match (case-insensitive)
  if (!match) {
    match = pages.find(p =>
      p.aliases?.some(a => a.toLowerCase() === lowerTarget)
    );
  }

  return match;
}

/**
 * Build the replacement wiki link for a dead link.
 *
 * @param page - The matching page
 * @param wikiFolder - Wiki root folder (e.g., "wiki")
 * @returns Formatted wiki link (e.g., "[[entities/chain-of-thought|Chain of Thought]]")
 *
 * @example
 * buildDeadLinkReplacement(
 *   { path: 'wiki/entities/chain-of-thought.md', title: 'Chain of Thought' },
 *   'wiki'
 * ) // => "[[entities/chain-of-thought|Chain of Thought]]"
 */
export function buildDeadLinkReplacement(
  page: PageRef,
  wikiFolder: string
): string {
  const relPath = page.path
    .replace(wikiFolder + '/', '')
    .replace('.md', '');
  return `[[${relPath}|${page.title}]]`;
}

/**
 * Replace dead link in content with corrected link.
 *
 * @param content - Source page content
 * @param targetName - Dead link target to find
 * @param replacement - Replacement wiki link
 * @returns Updated content with link replaced
 *
 * @example
 * replaceDeadLink('See [[思维链]] for details', '思维链', '[[entities/cot|Chain of Thought]]')
 * // => 'See [[entities/cot|Chain of Thought]] for details'
 */
export function replaceDeadLink(
  content: string,
  targetName: string,
  replacement: string
): string {
  const linkRegex = /\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]/g;
  return content.replace(
    linkRegex,
    (fullMatch: string, capturedTarget: string) => {
      if (capturedTarget.trim() === targetName) return replacement;
      return fullMatch;
    }
  );
}

File: python_lib/llm_wiki/wiki/lint/dead_link_detector.py
"""
Dead-link detector helpers — port of src/core/dead-link-detector.ts
"""
from __future__ import annotations

import re
from typing import Dict, List, Optional


def find_dead_link_target(pages: List[Dict], target_name: str) -> Optional[Dict]:
    """
    Find a matching page for a dead link target.
    Searches by title (case-insensitive) first, then by aliases.
    Each page dict should have 'path', 'title', and optionally 'aliases'.
    """
    target_basename = target_name.split("/")[-1] if "/" in target_name else target_name
    lower = target_basename.lower()

    match = next((p for p in pages if p.get("title", "").lower() == lower), None)
    if not match:
        match = next(
            (
                p for p in pages
                if any(a.lower() == lower for a in (p.get("aliases") or []))
            ),
            None,
        )
    return match


def build_dead_link_replacement(page: Dict, wiki_folder: str) -> str:
    """Build [[rel/path|Title]] replacement for a resolved dead link."""
    rel = page["path"].replace(wiki_folder + "/", "").replace(".md", "")
    return f"[[{rel}|{page['title']}]]"


def replace_dead_link(content: str, target_name: str, replacement: str) -> str:
    """Replace all [[targetName]] occurrences with the corrected link."""
    link_re = re.compile(r"\[\[([^\]|#]+)(?:[|#][^\]]+)?\]\]")

    def replacer(m: re.Match) -> str:
        if m.group(1).strip() == target_name:
            return replacement
        return m.group(0)

    return link_re.sub(replacer, content)
